A more detailed session for installation instructions (#2312)
CliMA/Oceananigans.jl: docs/src/installation_instructions.md

# Getting started

This section covers the basics of installing Julia and Oceananigans, and then describes best practices when running
Oceananigans. If you're new to Julia, we also recommend reading [Julia's "getting started" page](https://julialang.org/learning/getting-started/).

## Installation instructions

Usually the hardest part of installing Oceananigans is installing Julia. You can first check if your
system already has Julia installed. If that's not the case, installation is necessary and you can

> find instructions for your platform
> [here](https://julialang.org/downloads/platform/#linux_and_freebsd) and
> [here](https://julialang.org/downloads/). You can also manually compile Julia from source code
> following the instructions [here](https://github.com/JuliaLang/julia#building-julia).

Reviewer: ```suggestion
[download](https://julialang.org/downloads/) Julia from its official website, where also you can
find [platform-specific installation instructions](https://julialang.org/downloads/platform/).
Alternatively, you can also manually [compile Julia from source code](https://github.com/JuliaLang/julia#building-julia).
```

Reviewer 2: literate style <3

With Julia installed, you can install Oceananigans with a single line of code within Julia itself: `] add Oceananigans`. 
However, best practices for Julia suggest a couple extra commands to ensure that each project has
a separate and easily reproducible environment (we suggest that users read [this page](https://pkgdocs.julialang.org/v1/environments/)
to get a better understanding of working with environments). Best practices when starting a new Oceananigans 
project is to 

1. Start with a empty directory (in this example called `new_project`)
2. Inside the empty directory, start Julia with the command `julia --project`
3. Access the built-in package manager by pressing `]` in the Julia command prompt
4. Add the Oceananigans package with `add Oceananigans`
5. Instantiate/build all dependencies `instantiate`

To do so, open a Julia REPL from the terminal and then start the package manager by hitting `]`. Then:

```

> (@v1.6) pkg> activate .

Reviewer: there is an empty line before this which should be deleted

# Getting started

This section covers the basics of installing Julia and Oceananigans, and then describes best practices when running
Oceananigans. If you're new to Julia, we also recommend reading [Julia's "getting started" page](https://julialang.org/learning/getting-started/).

## Installation instructions

The hardest part of installing Oceananigans is actually installing Julia (and even that should be pretty easy). 
If you want to use Oceananigans on a laptop you can generally install Julia easily via the package manager of whatever
operational system you use (some quick googling should tell you how). If you're running on a cluster, you should check
if it already has Julia installed (most systems these days do). If that's the case, then the hardest part is already done!

If those options above are not possible, one alternative is to simply download the Julia pre-compiled binaries from the
[Julia website](https://julialang.org/downloads/), where you can simply run the Julia executable that's contained inside
(this should work out of the box!). Another option is to compile Julia manually from source code. This is definitely the hardest

> option, and you can find instructions on how to do that [here](https://github.com/JuliaLang/julia#building-julia).

Reviewer 2: This should be more concise. It's hard to skim these paragraphs and glean useful information. I don't think it's useful to say "installing Julia should be pretty easy". Users will determine for themselves whether its easy. We should have just 1-2 sentences about installing julia with relevant links I think.

Author: Done

[Julia website](https://julialang.org/downloads/), where you can simply run the Julia executable that's contained inside
(this should work out of the box!). Another option is to compile Julia manually from source code. This is definitely the hardest
option, and you can find instructions on how to do that [here](https://github.com/JuliaLang/julia#building-julia).

With Julia installed, you can install Oceananigans with a single line of code within Julia itself: `] add Oceananigans`. 
However, best practices for Julia suggest a couple extra commands to ensure that each project has
a separate and easily reproducible environment (we suggest that users read [this page](https://pkgdocs.julialang.org/v1/environments/)
to get a better understanding of working with environments). Best practices when starting a new Oceananigans 
project is to 

1. Start with a empty directory (in this example called `new_project`)
2. Inside the empty directory, start Julia with the command `julia --project`
3. Access the built-in package manager by pressing `]` in the Julia command prompt
4. Add the Oceananigans package with `add Oceananigans`
5. Instantiate/build all dependencies `instantiate`

> This process looks similar to this on a terminal (note that there may be many messages between some of these commands):

Reviewer: ```suggestion
To do so, open a Julia REPL from the terminal and then start the package manager by hitting `]`. Then:
```

```

> user@system:~$ julia --project
>                _
>    _       _ _(_)_     |  Documentation: https://docs.julialang.org
>   (_)     | (_) (_)    |
>    _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
>   | | | | | | |/ _` |  |
>   | | |_| | | | (_| |  |  Version 1.6.3 (2021-09-23)
>  _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
> |__/                   |

Reviewer: ```suggestion
```

(@v1.6) pkg> activate .

>   Activating new environment at `~/Dropbox/tests/new_project/Project.toml`

Reviewer: ```suggestion
  Activating new environment at `~/new_project/Project.toml`
```

```

At this time, updating should be done with care, as Oceananigans is under rapid development and breaking 
changes to the user API occur often. But if anything does happen, please open [an issue on github](https://github.com/CliMA/Oceananigans.jl/issues)!
We're more than happy to help with getting your simulations up and running.

!!! compat "Julia 1.6 or newer"
    The latest version of Oceananigans requires at least Julia v1.6 to run.
    Installing Oceananigans with an older version of Julia will install an older version of Oceananigans (the latest version compatible with your version of Julia).

## Running Oceananigans

Whenever you run Oceananigans for a project, it's recommended that you start Julia not with the simple command `julia`,
but with the command `julia --project`. This ensures that you will always use the package versions of you project
and that any changes there will not affect other projects. That said, running Oceananigans can be done a couple of ways. 

For initial explorations and getting used to the code, you can
enter the commands directly in the [Julia REPL](https://docs.julialang.org/en/v1/stdlib/REPL/) (which can be started with

> the command `julia --project`). This is an interactive

Reviewer: ```suggestion
the command `julia --project`). The REPL is an interactive
```

command-line tool (similar to Python's IPython) that has a searchable history, tab-completion, helpful keybindings, and many
other features to help users interact with packages. If you already have a file with commands, you can issue `julia --project my_oceananigans_script.jl`

> from the terminal, which will run all the commands in that script and subsequently close Julia. This is equivalent to opening

Reviewer: ```suggestion
from the terminal. This will execute all the commands in that script and subsequently close Julia. This is equivalent to opening
```

> the Julia REPL inside your project directory and typing each of those commands by hand (or simply issuing

Reviewer: ```suggestion
the Julia REPL inside your project directory and typing each of those commands by hand (or simply calling
```

> `include("my_oceananigans_script.jl")` in the REPL), with the exception that the REPL will remain open after the commands in this case.

Reviewer: ```suggestion
`include("my_oceananigans_script.jl")` in the REPL).
```

## Running Oceananigans

Whenever you run Oceananigans for a project, it's recommended that you start Julia not with the simple command `julia`,
but with the command `julia --project`. This ensures that you will always use the package versions of you project
and that any changes there will not affect other projects. That said, running Oceananigans can be done a couple of ways. 

For initial explorations and getting used to the code, you can
enter the commands directly in the [Julia REPL](https://docs.julialang.org/en/v1/stdlib/REPL/) (which can be started with
the command `julia --project`). This is an interactive
command-line tool (similar to Python's IPython) that has a searchable history, tab-completion, helpful keybindings, and many
other features to help users interact with packages. If you already have a file with commands, you can issue `julia --project my_oceananigans_script.jl`
from the terminal, which will run all the commands in that script and subsequently close Julia. This is equivalent to opening
the Julia REPL inside your project directory and typing each of those commands by hand (or simply issuing

> `include("my_oceananigans_script.jl` in the REPL), with the exception that the REPL will remain open after the commands in this case.

Reviewer: ```suggestion
`include("my_oceananigans_script.jl")` in the REPL), with the exception that the REPL will remain open after the commands in this case.
```

If you are using a HPC system and you want to make use of a GPU, then an extra step generally has to be added: submitting the
job to request GPU resources. This is usually done with the `slurm` or `PBS` job schedulers and the details usually change depending
on which cluster you are using, but the process should be similar (if not the same) to submitting other jobs in that same cluster. We

> recommend you check the cluster documentation for information on how to submit jobs that require GPU resources.

Reviewer: This has nothing to do with Oceananigans.

Author: True, which is why there's a referral to whatever HPC system the user is using. But I'd argue that's useful. Having that part there would definitely have made my life easier 2 years ago when I started using Oceananigans.

Although I agree that we might wanna tie that with the act of "running Oceananigans" a bit more

Author: @navidcy check this newer version which is more brief. I think that's useful for new users, but if you still feel like that's out of place, then I'm okay removing it.

Reviewer 2: I tend to agree that we shouldn't put info here that's not relevant to Oceananigans. There's an infinite number of things that might make all our lives easier...

Reviewer 2: Perhaps more to the point, if there is a need for "tips for using GPUs with slurm and PBS (and Julia?)" then that information belongs elsewhere. It's _less_ helpful to all those suffering humans to bury this info in the Oceananigans docs!

Reviewer: Yes. Let's drop this bit.

## Installation instructions

Usually the hardest part of installing Oceananigans is installing Julia. The main options here in order
of simple to complex are

> 1. Check if your system already has Julia installed. Most HPC systems these days have Julia

Reviewer 3: Should we add a sentence like?: 

If you are using an HPC, run `module spider julia` to check the available versions. For now, `Oceananigans` is only supported by `Julia-1.6.5 (LTS)`.

Author: I don't know if `module spider` is common to all HPC systems, so I'm reluctant to include that info there

Reviewer 3: I also don't know. haha
Well, the clusters I have used so far have different jobs managers and still works with `module avail` or `module spider`.
What do you think @glwagner and @navidcy ?

Reviewer 2: I think its not right for the docs which should focus on documenting the code. I think it'd be great to have helpful info and tips like that somewhere for users. A discussion might be a great place because everyone can contribute and its more living (we don't need PRs to update it)

Reviewer: I agree with @glwagner. The docs should make clear which version of Julia Oceananigans.jl require (which they do). We cannot document in Oceananigans docs details about all HPC systems.

>    available.

Reviewer 2: Are these instructions only for HPC?

Author: Not necessarily. I wouldn't discard the possibility of some OS coming with Julia pre-installed (although I don't know of any of this point).

> 2. Install Julia via the package manager of the operational system you use. E.g. `apt` for Ubuntu.

Reviewer 2: I don't think Julia recommends this, so we shouldn't, right? I don't see it mentioned here:

https://julialang.org/downloads/platform/#linux_and_freebsd

There is however an installer for windows:

https://julialang.org/downloads/platform/#juliaup

other projects.. To activate the project's environment you either start Julia using `julia --project` or from Julia's
package manager you call:
```Julia

> (@v1.6) pkg> activate .

Reviewer: oops! :)
